### PR TITLE
Add benchmark for String.indexOf(String)

### PR DESCRIPTION
**Pull request by NigelYiboYu**
Add benchmark for String.indexOf(String) API

Signed-off-by: Nigel Yu <yunigel@ca.ibm.com>

### CODE REVIEW

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+*******************************************************************************/
+
+package net.adoptopenjdk.bumblebench.string;
+
+import java.util.Random;
+
+import net.adoptopenjdk.bumblebench.core.MicroBench;
+
+/**
+ *
```
Comment (fjeremic) [inline net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Some extra new lines here not needed.

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+	protected long doBatch(long numLoops) throws InterruptedException {
+
+		for (long loop = 0; loop < numLoops; loop++) {
+			startTimer();
+			int total = doIndexOfString();
+			pauseTimer();
+			_escape = total;
+
+			// verify
```
Comment (fjeremic) [inline net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Unnecessary comment.

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ * 
+ * Use -DBumbleBench.stringLength=1 to 4 and -DBumbleBench.indexLevel=1 to 4 too
+ * choose testing modes.
+ */
+public class StringIndexOfStringBench extends MicroBench {
+
+	static final int MAX_NUM_STRINGS = option("maxNumStrings", 5);
+	static int stringLength = option("stringLength", 1);
+	static int indexLevel = option("indexLevel", 1);
+	static boolean verifyResult = option("verify", 0) != 0;
```
Comment (fjeremic) [inline net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Just confirming `verifyResult == false` by default?

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	protected long doBatch(long numLoops) throws InterruptedException {
+
+		for (long loop = 0; loop < numLoops; loop++) {
+			startTimer();
+			int total = doIndexOfString();
+			pauseTimer();
+			_escape = total;
+
+			// verify
+			if (verifyResult) {
```
Comment (fjeremic) [inline net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: I suppose it doesn't hurt, but this is a benchmark. We should be assuming the results of the operations are correct. Functional testing should take care that the `indexOf` operation gives correct results. I'm not sure this code belongs in a performance benchmark.

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java (hunk 1 above)
Comment (NigelYiboYu) [reply net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Fixed

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java (hunk 2 above)
Comment (NigelYiboYu) [reply net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Fixed

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java (hunk 4 above)
Comment (NigelYiboYu) [reply net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: Removed all result verification logic.

File: net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java (hunk 3 above)
Comment (NigelYiboYu) [reply net/adoptopenjdk/bumblebench/string/StringIndexOfStringBench.java]: This has been removed.

**Review by fjeremic — APPROVED** [review]

**Comment by NigelYiboYu** [comment]
Updated the benchmark to use a human-readable string as its default source string, based on which sub-strings are generated. This 64K character long source string is preprocessed to contain only compressible characters.

**Review by fjeremic — APPROVED** [review]

**Comment by fjeremic** [comment]
@piyush286 / @smlambert do you might know how typically gets assigned to be a committer for these PRs? It would still be valuable to have them merged.

**Comment by piyush286** [comment]
I usually request a review from @smlambert, who can throw more light on this.

**Comment by smlambert** [comment]
Did not see this as active, so it fell off my radar, apologies. 

I can review, and would merge once I run a Grinder (assuming it will get exercised as part of this target: https://github.com/AdoptOpenJDK/openjdk-tests/blob/master/perf/bumbleBench/playlist.xml#L539).

**Comment by piyush286** [comment]
This is a new test case `StringIndexOfStringBench` so we'll need to add it to our BumbleBench playlist file. I can add it and test this change once we're done with Quarkus testing, unless someone else gets to it before that.

The above test case is for `StringIndexOfBench`:
https://github.com/AdoptOpenJDK/bumblebench/blob/master/net/adoptopenjdk/bumblebench/string/StringIndexOfBench.java
https://github.com/AdoptOpenJDK/openjdk-tests/blob/dc1569c25d402cd3c4936f95ed49bcf5786b1ef9/perf/bumbleBench/playlist.xml#L540

**Review by smlambert — APPROVED** [review]

**Comment by smlambert** [comment]
Run locally, we can/should create a new test target for it next week.

**Comment by fjeremic** [comment]
Thanks everyone!

**Pull request merged**